feat: enable job logs command

bvtujo opened the pull request:
- chore: update cloudwatchlogs to allow limiting log streams fetched and custom prefixes
- chore: add logic for fetching state machine logs to workload
- chore: add unit tests for workload logging pkg
- chore: make things work for ci
- chore: address feedback
- feat: add job logs logic
- chore: add logic in underlying libraries for job logs (#3757)
- fix logic bugs in log stream sorting
- chore: make IncludeStateMachine a logging option, not a struct member
- feat: enable job logs command and add tests

Adds job logs command and fixes a bug which randomized the sort order of log streams.

<!-- Provide summary of changes -->

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, 77" -->

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

bvtujo committed:
chore: update cloudwatchlogs to allow limiting log streams fetched and custom prefixes

bvtujo committed:
chore: add logic for fetching state machine logs to workload

bvtujo committed:
chore: add unit tests for workload logging pkg

bvtujo committed:
chore: make things work for ci

bvtujo committed:
feat: add job logs logic

feat: enable command

test: add job logs ask and validate tests; rename svcAppNameHelpPrompt

chore: add execute tests

bvtujo committed:
chore: address feedback

use pagination for log stream fetching.
rename variables.
make sure log stream truncation and filtering are engaged in tests.

bvtujo committed:
chore: add logic in underlying libraries for job logs (#3757)

- chore: add job logs logic and validation
- chore: update cloudwatchlogs to allow limiting log streams fetched and custom prefixes
- chore: rename service.go to workload.go in logging pkg
- chore: add logic for fetching state machine logs to workload
- chore: add unit tests for workload logging pkg

Adds unit tests and changes for the cloudwatchlogs and logging libraries necessary for job logs, specifically for getting one or N executions at a time, optionally including state machine logs.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

bvtujo committed:
fix logic bugs in log stream sorting

bvtujo committed:
chore: make IncludeStateMachine a logging option, not a struct member

bvtujo committed:
feat: enable job logs command and add tests

Lou1415926 commented on internal/pkg/aws/cloudwatchlogs/cloudwatchlogs.go:
```diff
@@ -185,16 +195,13 @@ func initGetLogEventsInput(opts LogEventsOpts) *cloudwatchlogs.GetLogEventsInput
 		for _, prefix := range prefixes {
 			if strings.HasPrefix(candidate, prefix) {
-				m[candidate] = true
+				res = append(res, candidate)
```
I see this is getting sort of addressed in #3773!

iamhopaul123 replied:
This function is updated now so that we can deal with the matching more efficiently.

Lou1415926 commented on internal/pkg/cli/job_logs.go:
```diff
@@ -263,7 +278,13 @@ Displays logs from specific task IDs.
 	cmd.Flags().BoolVar(&vars.follow, followFlag, false, followFlagDescription)
 	cmd.Flags().DurationVar(&vars.since, sinceFlag, 0, sinceFlagDescription)
 	cmd.Flags().IntVar(&vars.limit, limitFlag, 0, limitFlagDescription)
+	cmd.Flags().IntVar(&vars.last, lastFlag, 6, lastFlagDescription)
```
why do we want 6 for the default?

Lou1415926 commented on internal/pkg/cli/job_logs.go:
```diff
@@ -151,38 +157,43 @@ func (o *jobLogsOpts) Ask() error {
+		StartTime:               o.startTime,
+		TaskIDs:                 o.taskIDs,
+		OnEvents:                eventsWriter,
+		LogStreamLimit:          logStreamLimit,
```
I realized I've never noticed this parameter existed before :homer-disappearing !  If `logStreamLimit`, is it guaranteed that the one and only one log stream that gets fetched is the log stream that's produced by the latest call? Could it be the log stream that's produced by the second lastest call? (sorry for the repetition of words 😂 )

bvtujo replied:
It's the most recent that Cloudwatch has to offer; I don't know that we can control the delay between a container coming up and the log stream for that container appearing. But I assume that there will be few times when someone needs to immediately view the logs from the last execution less than 10 seconds after the job is triggered, and this case can be covered by
```sh
copilot job run && copilot job logs --follow
``` 
or eventually
```sh
copilot job run --follow
```

bvtujo committed:
docs: add job logs docs

bvtujo committed:
chore: change default number of executions to show

rclinard-amzn commented on site/content/docs/commands/job-logs.en.md:
```diff
@@ -0,0 +1,62 @@
+
+## What does it do?
+
+`copilot job logs` displays the logs of a deployed service.
```
This might need to be changed in the code as well
```suggestion
`copilot job logs` displays the logs of a deployed job.
```

rclinard-amzn commented on site/content/docs/commands/job-logs.en.md:
```diff
@@ -0,0 +1,62 @@
+                                logs should be shown. (default 3)
+      --limit int               Optional. The maximum number of log events returned. Default is 10
+                                unless any time filtering flags are set.
+  -n, --name string             Name of the service.
```
This might need to be changed in the code as well
```suggestion
  -n, --name string             Name of the job.
```

rclinard-amzn commented on site/content/docs/commands/job-logs.en.md:
```diff
@@ -0,0 +1,62 @@
+      --json                    Optional. Output in JSON format.
+      --last int                Optional. The number of executions of the scheduled job for which
+                                logs should be shown. (default 3)
+      --limit int               Optional. The maximum number of log events returned. Default is 10
+                                unless any time filtering flags are set.
```
Out of curiosity, is the default actually 10? It seems to be set to 0 in the code but I haven't read the code close enough to figure out whether it gets defaulted to 10 later.

iamhopaul123 replied:
This is the one https://github.com/aws/copilot-cli/blob/e1083b91cd96a0c3123789ac8638b6e181540ffe/internal/pkg/logging/workload.go#L23

iamhopaul123 commented on internal/pkg/cli/cli.go:
```diff
@@ -23,8 +23,8 @@ import (
 
 const (
-	svcAppNamePrompt     = "Which application does your service belong to?"
-	svcAppNameHelpPrompt = "An application groups all of your services and jobs together."
```
🤦

iamhopaul123 commented on internal/pkg/aws/cloudwatchlogs/cloudwatchlogs.go:
```diff
@@ -101,15 +101,16 @@ func (c *CloudWatchLogs) logStreams(logGroup string, logStreamLimit int, logStre
 
-	return truncateStreams(logStreamLimit, logStreamNames), nil
+	// Make sure that the streams are returned with the latest events last so they're printed to terminal in chronological order.
+	return reverseStreams(logStreamNames), nil
```
Hmm I think I would expect last events to be first for logs 🤔

bvtujo replied:
I wrote a long digression on why this is necessary, but I guess though it actually doesn't matter. Later on, in `LogEvents()`, we sort all the logs by event time so the order of streams doesn't matter, just that they are the most recent N streams chronologically.

iamhopaul123 commented on internal/pkg/cli/job_logs.go:
```diff
@@ -151,38 +157,43 @@ func (o *jobLogsOpts) Ask() error {
-		if _, err := o.configStore.GetJob(o.appName, o.name); err != nil {
-			return err
-		}
+	if o.includeStateMachineLogs {
+		logStreamLimit *= 2
```
Is this a safe assumption that we can make 💭 If so we might need to add some comments explaining the reason. Also I feel like we should deal with the logic within the logging pkg, because otherwise we are passing duplicated info around.

bvtujo replied:
Good idea. I guess we can use more jobby abstractions in `logging` since it's not the low-level aws package.

bvtujo committed:
nit: don't reverse log stream order

It's made moot later, see this comment:
https://github.com/aws/copilot-cli/pull/3794/files#r933412494

bvtujo committed:
chore: move state machine logic into logging package

bvtujo committed:
chore: fix tests

bvtujo committed:
docs: fix nits in docs

bvtujo committed:
Merge branch 'mainline' into job-logs

Lou1415926 commented on site/content/docs/concepts/jobs.en.md:
```diff
@@ -41,4 +41,18 @@ Since Copilot uses CloudFormation under the hood, all the resources created are
\ No newline at end of file
+### Where are my job's logs?
+
+Checking your job logs is easy as well. Running `copilot job logs` will show the most recent logs of your job. You can follow your logs live with the `--follow` flag,
```
nit:
```suggestion
Checking your job logs is easy as well. Running [`copilot job logs`](../commands/job-log.en.md) will show the most recent logs of your job. You can follow your logs live with the `--follow` flag,
```

Lou1415926 commented on internal/pkg/cli/job_logs.go:
```diff
@@ -198,54 +205,58 @@ func (o *jobLogsOpts) getTargetEnv() (*config.Environment, error) {
+  Displays logs from the last execution of the job.
+  /code $ copilot job logs --last 1
+  Displays logs from specific task IDs.
   /code $ copilot job logs --tasks 709c7eae05f947f6861b150372ddc443,1de57fd63c6a4920ac416d02add891b9
```
nit: maybe we can update this example as 
```suggestion
/code $ copilot job logs --tasks 709c7ea,1de57fd
```
I took this from the doc - it looks cleaner and is consistent with the doc!

Lou1415926 commented on site/content/docs/commands/job-logs.en.md:
````diff
@@ -0,0 +1,62 @@
+```console
+$ copilot job logs --tasks 709c7ea,1de57fd
+```
+
````
```suggestion
Displays logs in real time.
```console
$ copilot job logs --follow
```


```

Lou1415926 commented on internal/pkg/logging/workload_test.go:
```diff
@@ -165,12 +165,13 @@ firelens_log_router/fcfe4 10.0.0.00 - - [01/Jan/1970 01:01:01] "WARN some warnin
 				gomock.InOrder(
 					m.logGetter.EXPECT().LogEvents(gomock.Any()).
 						Do(func(param cloudwatchlogs.LogEventsOpts) {
 							require.Equal(t, param.LogStreamPrefixFilters, []string{"copilot/", "states"})
-							require.Equal(t, param.Limit, aws.Int64(10))
+							require.Equal(t, param.LogStreamLimit, 2)
```
are we still interested in what's passed in `param.Limit`? If so, what do you think of
```go
require.Equal(t, param.LogStreamPrefixFilters, []string{"copilot/", "states"})
require.Equal(t, param,Limit aws.Int(10))
require.Equal(t, param.LogStreamLimit, 2)
```

bvtujo replied:
Good idea; i've added a check here and a new test to ensure that some log stream limit behavior is working properly.

bvtujo committed:
Update site/content/docs/concepts/jobs.en.md

Co-authored-by: Wanxian Yang <79273084+Lou1415926@users.noreply.github.com>

bvtujo committed:
Update internal/pkg/cli/job_logs.go

Co-authored-by: Wanxian Yang <79273084+Lou1415926@users.noreply.github.com>

bvtujo committed:
Update site/content/docs/commands/job-logs.en.md

Co-authored-by: Wanxian Yang <79273084+Lou1415926@users.noreply.github.com>

bvtujo committed:
test: add additional param checks and unit test to validate limit and logStreamLimit work together

Lou1415926 reviewed (approved):
LGTM! Feel free to remove the label when you are ready (I did it but then added it back just in case you wanted to add anything)

mergify[bot] committed:
Merge branch 'mainline' into job-logs
